### DISCUSSION TITLE
Custom control based on ListBox doesn't show any content

### DISCUSSION THREAD

**Discussion by radaiko** [AvaloniaUI/Avalonia#16248] [Q&A] [answered]
This is my custom ListBox with as minimal code as possible:
```csharp
public class OrderAbleListBoxTaskItem : ListBox {
  public ObservableCollection<TaskItem>? AllItems { get; init; }
  
  public OrderAbleListBoxTaskItem() : base() {
    ItemsSource = AllItems;
    Height = 300;
    Width = 300;
  }
}
```

When I try to use this custom ListBox the items are filled in the debugger but nothing is displayed.
When I try to use the default ListBox the items are shown.

This is my code I used for testing what it shown:
```csharp
// DUMMY
var listbox = new OrderAbleListBoxTaskItem {
  AllItems = (ObservableCollection<TaskItem>?)_tasks.GetAll()
};
TaskList.Children.Add(listbox);
TaskList.Children.Add(new ListBox() { ItemsSource = _tasks.GetAll() });
```

As we can see here on the top is the 300x300 custom ListBox completely empty and on the bottom the default on:
![image](https://github.com/AvaloniaUI/Avalonia/assets/15850003/d7523a93-8ad7-4521-8ca7-80cd70e9a906)

What I'm doing wrong?

**Comment by mikegoatly** [comment]
The order your code will be executed in is:

* _Constructor_: Sets `ItemsSource` to `AllItems` (currently null)
* _Object initialization_: Sets `AllItems` to `(ObservableCollection<TaskItem>?)_tasks.GetAll()`

Because you're not updating the ListBox's `ItemsSource` after `AllItems` is set, the list is going to remain empty.

I think if you need to respond to a property changing on a control like this, you'll need to use a StyledProperty or DirectProperty and override `OnPropertyChanged`.

**Reply by radaiko** [reply]
Even if I prefill AllItems, like this, it still will not show anything.

```csharp
public class OrderAbleListBoxTaskItem : ListBox {
  public ObservableCollection<TaskItem> AllItems { get; init; } = [];
  
  public OrderAbleListBoxTaskItem() {
    AllItems.Add(new TaskItem() {Summary = "t1"});
    AllItems.Add(new TaskItem() {Summary = "t2"});
    AllItems.Add(new TaskItem() {Summary = "t3"});
    AllItems.Add(new TaskItem() {Summary = "t4"});
    ItemsSource = AllItems;
    Height = 300;
    Width = 300;
  }
}
```

**Comment by stevemonaco** [comment, marked as answer]
This should be because your custom control is not receiving a `ControlTheme`. Therefore it has no appearance and nothing to draw. You need to create a `ControlTheme` or use `StyleKeyOverride`. See https://docs.avaloniaui.net/docs/basics/user-interface/styling/styles#style-key if you want it to receive the `ListBox` style.

**Reply by radaiko** [reply]
Worked perfect. Thanks.
